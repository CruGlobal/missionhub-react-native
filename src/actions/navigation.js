import { NavigationActions } from 'react-navigation';

// Performance issues: https://github.com/react-community/react-navigation/issues/608#issuecomment-328635042

export function navigatePush(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.navigate({
      routeName: screen,
      params: props,
    }));
  };
}

export function navigateBack(times, backParams) {
  return (dispatch) => {
    if (times && times > 1) {
      dispatch(NavigationActions.pop({ n: times, immediate: true }));
    } else {
      // backParams can contain { key: string, immediate: bool }
      dispatch(NavigationActions.back(backParams));
    }
  };
}

export function navigateReset(screen, props = {}) {
  return (dispatch) => {
    dispatch(NavigationActions.reset({
      index: 0,
      actions: [
        NavigationActions.navigate({ routeName: screen, params: props }),
      ],
    }));
  };
}

// The reset home and reset login are handled by the login/logout auth actions
