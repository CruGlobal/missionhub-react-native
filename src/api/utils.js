import merge from 'lodash/merge';
import qs from 'qs';

const API_VERSION = 'v4';
let baseUrl = '';
if (__DEV__) {
  baseUrl = 'https://api-stage.missionhub.com';
} else {
  baseUrl = 'https://api.missionhub.com';
}

export const BASE_URL = baseUrl;
export const API_URL = `${BASE_URL}/apis/${API_VERSION}/`;
export const THE_KEY_URL = 'https://thekey.me/cas/';

const DEFAULT_HEADERS = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
};

const ERROR_CODES = [ 400, 401, 402, 403, 404, 500, 504 ];
export function handleResponse(response) {
  if (!response) return null;
  if (response && ERROR_CODES.includes(response.status)) {
    return response.json().then((jsonResponse) => Promise.reject(jsonResponse));

  } else {
    return json(response);
    // return response;
  }
}

export function json(response) {
  return response.text().then((t) => t ? JSON.parse(t) : null);
}

function createUrl(url = '', params) {
  let newUrl = `${url}`;
  if (newUrl[0] === '/') {
    newUrl = newUrl.substr(1);
  }
  let fullUrl = newUrl;
  if (params && Object.keys(params).length > 0) {
    let paramsStr = qs.stringify(params);
    // let paramsStr = Object.keys(params).map((p) => `${p}=${params[p]}`).join('&');
    // if (paramsStr) {
    // fullUrl += `?${paramsStr}`;
    // }
    fullUrl += `?${paramsStr}`;
  }
  return fullUrl;
}

function defaultObject(method, obj = {}, data) {
  let newObj = merge({}, { headers: DEFAULT_HEADERS }, obj, { method: method.toUpperCase() });
  if (obj && obj.headers && obj.headers['Content-Type'] === false) {
    delete newObj.headers['Content-Type'];
  }
  if (obj && obj.headers && obj.headers['Accept'] === false) {
    delete newObj.headers['Accept'];
  }
  if (data) {
    newObj.body = obj.stringify === false ? data : JSON.stringify(data);
  }
  delete obj.stringify;

  return newObj;
}


export default function request(type, url, query, data, extra) {
  const newUrl = createUrl(url, query);
  const newObject = defaultObject(type, extra, data);
  APILOG('REQUEST', newObject.method, newUrl, newObject);

  return fetch(newUrl, newObject).then(handleResponse);
}
