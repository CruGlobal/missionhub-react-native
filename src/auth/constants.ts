import { AuthPerson_currentUser_person } from './__generated__/AuthPerson';

export enum IdentityProvider {
  Apple,
  Google,
  Facebook,
  TheKey,
  Anonymous,
}

export enum AuthError {
  None,
  CredentialsIncorrect,
  EmailUnverified,
  MfaRequired,
  MfaIncorrect,
  Unknown,
}

export const emptyAuthPerson: AuthPerson_currentUser_person = {
  __typename: 'Person' as const,
  id: '',
  firstName: '',
  lastName: '',
  fullName: '',
  stage: null,
  globalRegistryMdmId: null,
  fbUid: null,
  theKeyUid: null,
};
