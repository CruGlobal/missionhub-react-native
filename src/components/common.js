// Common components that we can import with simple destructuring syntax
// So when one of these components are loaded in, all files will be called also
import Button from './Button';
import DateComponent from './DateComponent';
import Flex from './Flex';
import Icon from './Icon';
import Input from './Input';
import Loading from './Loading';
import Text from './Text';
import Touchable from './Touchable';
import Separator from './Separator';
import RefreshControl from './RefreshControl';

export {
  Button,
  DateComponent,
  Flex,
  Icon,
  Input,
  Loading,
  Text,
  Touchable,
  Separator,
  RefreshControl,
};
