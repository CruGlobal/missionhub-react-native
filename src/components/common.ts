// Common components that we can import with simple destructuring syntax
// So when one of these components are loaded in, all files will be called also
import Button from './Button';
import Card from './Card';
import DateComponent from './DateComponent';
import Dot from './Dot';
import Flex from './Flex';
import Icon from './Icon';
import IconButton from './IconButton';
import Input from './Input';
import LoadingGuy from './LoadingGuy';
import LoadingWheel from './LoadingWheel';
import PlatformKeyboardAvoidingView from './PlatformKeyboardAvoidingView';
import RadioButton from './RadioButton';
import RefreshControl from './RefreshControl';
import Separator from './Separator';
import Text from './Text';
// eslint-disable-next-line import/no-unresolved
// @ts-ignore
import Touchable from './Touchable';

export {
  Button,
  Card,
  DateComponent,
  Dot,
  Flex,
  Icon,
  IconButton,
  Input,
  LoadingGuy,
  LoadingWheel,
  PlatformKeyboardAvoidingView,
  RadioButton,
  RefreshControl,
  Separator,
  Text,
  Touchable,
};
