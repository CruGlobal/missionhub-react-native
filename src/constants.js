export const LOGIN = 'app/LOGIN';
export const LOGOUT = 'app/LOGOUT';
export const FIRST_TIME = 'app/FIRST_TIME';
export const STAGES = 'app/STAGES';
export const SELECT_STAGE = 'app/SELECT_STAGE';
export const FIRST_NAME_CHANGED = 'app/FIRST_NAME_CHANGED';
export const LAST_NAME_CHANGED = 'app/LAST_NAME_CHANGED';
export const SAVE_NOTES = 'app/SAVE_NOTES';
export const PERSON_FIRST_NAME_CHANGED = 'app/PERSON_FIRST_NAME_CHANGED';
export const PERSON_LAST_NAME_CHANGED = 'app/PERSON_LAST_NAME_CHANGED';
export const ADD_STEP_REMINDER = 'app/ADD_STEP_REMINDER';
export const REMOVE_STEP_REMINDER = 'app/REMOVE_STEP_REMINDER';
export const PUSH_NOTIFICATION_SHOULD_ASK = 'app/PUSH_NOTIFICATION_SHOULD_ASK';
export const PUSH_NOTIFICATION_ASKED = 'app/PUSH_NOTIFICATION_ASKED';
export const PUSH_NOTIFICATION_SET_TOKEN = 'app/PUSH_NOTIFICATION_SET_TOKEN';
export const PUSH_NOTIFICATION_REMINDER = 'app/PUSH_NOTIFICATION_REMINDER';
export const PEOPLE_WITH_ORG_SECTIONS = 'app/PEOPLE_WITH_ORG_SECTIONS';
export const SWIPE_REMINDER_STEPS_HOME = 'app/SWIPE_REMINDER_STEPS_HOME';
export const SWIPE_REMINDER_STEPS_CONTACT = 'app/SWIPE_REMINDER_STEPS_CONTACT';
export const SWIPE_REMINDER_STEPS_REMINDER = 'app/SWIPE_REMINDER_STEPS_REMINDER';
export const SWIPE_REMINDER_JOURNEY = 'app/SWIPE_REMINDER_JOURNEY';
export const SET_VISIBLE_PERSON_INFO = 'app/SET_VISIBLE_PERSON_INFO';
export const UPDATE_VISIBLE_PERSON_INFO = 'app/UPDATE_VISIBLE_PERSON_INFO';
export const SET_JOURNEY_ITEMS = 'app/SET_JOURNEY_ITEMS';
export const CLEAR_JOURNEY_ITEMS = 'app/CLEAR_JOURNEY_ITEMS';
export const COMPLETED_STEP_COUNT = 'app/COMPLETED_STEP_COUNT';
export const UPDATE_STAGES = 'app/UPDATE_STAGES';
export const UPDATE_ONBOARDING_PERSON = 'app/UPDATE_ONBOARDING_PERSON';
export const RESET_ONBOARDING_PERSON = 'app/RESET_ONBOARDING_PERSON';

export const STEP_NOTE = 'stepNote';

export const NAVIGATE_FORWARD = 'Navigation/NAVIGATE';
export const NAVIGATE_RESET = 'Navigation/RESET';
export const MAIN_TABS = 'nav/MAIN_TABS';
export const DRAWER_OPEN = 'DrawerOpen';
export const DRAWER_CLOSE = 'DrawerClose';
export const MAIN_MENU_DRAWER = 'nav/drawer/main';
export const CONTACT_MENU_DRAWER = 'nav/drawer/contact';

export const EXPIRED_ACCESS_TOKEN = 'Expired access token';

export const URL_ENCODED = 'application/x-www-form-urlencoded';
export const THE_KEY_CLIENT_ID = '8480288430352167964';

export const CASEY = 'casey';
export const JEAN = 'jean';

export const LINKS = {
  about: 'https://get.missionhub.com',
  help: 'http://help.missionhub.com',
  playStore: 'market://details?id=com.missionhub',
  appleStore: 'itms://itunes.apple.com/us/app/apple-store/id447869440?mt=8',
  terms: 'https://get.missionhub.com/terms-of-service/',
};

export const ANALYTICS_CONTEXT_CHANGED = 'app/ANALYTICS_CONTEXT_CHANGED';

export const ORG_PERMISSIONS = [ 1, 4 ];

export const ANALYTICS = {
  PAGE_NAME: 'PageName',
  MCID: 'cru.mcid',
  SCREENNAME: 'cru.screenname',
  PREVIOUS_SCREENNAME: 'cru.previousscreenname',
  SITE_SECTION: 'cru.sitesection',
  SITE_SUBSECTION: 'cru.sitesubsection',
  SITE_SUB_SECTION_3: 'cru.subsectionlevel3',
  CONTENT_AUDIENCE_TARGET: 'cru.contentaudiencetarget',
  CONTENT_TOPIC: 'cru.contenttopic',
  LOGGED_IN_STATUS: 'cru.loggedinstatus',
  SSO_GUID: 'cru.ssoguid',
  GR_MASTER_PERSON_ID: 'cru.grmasterpersonid',
  FACEBOOK_ID: 'cru.facebookid',
  CONTENT_LANGUAGE: 'cru.contentlanguage',
};
export const LOGGED_IN = 'logged in';
export const NOT_LOGGED_IN = 'not logged in';

export const INTERACTION_TYPES = {
  MHInteractionTypeAssignedContacts: { id: 100, requestFieldName: 'contact_count', iconName: 'peopleIcon', translationKey: 'interactionAssignedContacts' },
  MHInteractionTypeUncontacted: { id: 101, requestFieldName: 'uncontacted_count', iconName: 'spiritualConversationsIcon', translationKey: 'interactionUncontacted' },
  MHInteractionTypeNote: { id: 1, iconName: 'commentIcon', translationKey: 'interactionNote', hideReport: true },
  MHInteractionTypeSpiritualConversation: { id: 2, iconName: 'spiritualConversationsIcon', translationKey: 'interactionSpiritualConversation' },
  MHInteractionTypeGospelPresentation: { id: 3, iconName: 'gospelIcon', translationKey: 'interactionGospel' },
  MHInteractionTypePersonalDecision: { id: 4, iconName: 'decisionIcon', translationKey: 'interactionDecision' },
  MHInteractionTypeHolySpiritConversation: { id: 5, iconName: 'spiritIcon', translationKey: 'interactionSpirit' },
  // MHInteractionTypeGraduatingOnMission: { id: 6, iconName: '', translationKey: 'interactionNote' },
  MHInteractionTypeDiscipleshipConversation: { id: 9, iconName: 'discipleshipConversationIcon', translationKey: 'interactionDiscipleshipConversation' },
};

export default {

};
