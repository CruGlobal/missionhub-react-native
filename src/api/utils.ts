import qs from 'qs';
import Config from 'react-native-config';

import { APILOG } from '../utils/logging';

const API_VERSION = 'v4';

export const BASE_URL = Config.API_BASE_URL;
export const API_URL = `${BASE_URL}/apis/${API_VERSION}`;
export const THE_KEY_URL = Config.THE_KEY_URL;

const DEFAULT_HEADERS = {
  Accept: 'application/json',
  'Content-Type': 'application/json',
};

// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
const ERROR_CODES = [
  400,
  401,
  402,
  403,
  404,
  408,
  413,
  422,
  429,
  431,
  500,
  501,
  502,
  503,
  504,
];
const JSON_CONTENT_TYPE = ['application/json', 'application/vnd.api+json'];

export type ApiErrorResponse =
  | {
      errors?: { detail?: string }[];
      error?: string;
    }
  | string;

async function handleResponse(response: Response) {
  if (!response) {
    return { jsonResponse: null, sessionHeader: '' };
  }

  if (response && ERROR_CODES.includes(response.status)) {
    // Check the content type and try to parse the error correctly based on the type
    const contentType = response.headers.get('content-type') || '';
    if (
      contentType.indexOf(JSON_CONTENT_TYPE[0]) !== -1 ||
      contentType.indexOf(JSON_CONTENT_TYPE[1]) !== -1
    ) {
      return response.json().then((r: ApiErrorResponse) => Promise.reject(r));
    }
    return response.text().then((r: ApiErrorResponse) => Promise.reject(r));
  }

  // We need to use response.text() and JSON.parse(t) for responses because we are getting data back with
  // a content-type of 'application/vnd.api+json' which needs to be handled like this
  const text = await response.text();
  return {
    jsonResponse: text ? JSON.parse(text) : null,
    sessionHeader: response.headers.get('X-MH-Session'),
  };
}

function createUrl(url = '', params: Record<string, unknown>) {
  let newUrl = `${url}`;
  if (newUrl[0] === '/') {
    newUrl = newUrl.substr(1);
  }
  let fullUrl = newUrl;
  if (params && Object.keys(params).length > 0) {
    // stringify arrays into brackets without indices https://github.com/ljharb/qs#stringifying
    const paramsStr = qs.stringify(params, { arrayFormat: 'brackets' });

    fullUrl += `?${paramsStr}`;
  }
  return fullUrl;
}

const defaultObject = (
  method: string,
  extra: RequestInit,
  data: Record<string, unknown> | BodyInit | undefined,
  stringify = true,
  // eslint-disable-next-line max-params
): RequestInit => ({
  ...extra,
  headers: { ...DEFAULT_HEADERS, ...(extra.headers ? extra.headers : {}) },
  method: method.toUpperCase(),
  ...(data
    ? {
        body: stringify === false ? (data as BodyInit) : JSON.stringify(data),
      }
    : {}),
});

// eslint-disable-next-line max-params
export default async function request(
  type: string,
  url: string,
  query: Record<string, unknown>,
  data: Record<string, unknown> | undefined,
  extra: RequestInit,
  stringify = true,
) {
  const newUrl = createUrl(url, query);
  const newObject = defaultObject(type, extra, data, stringify);
  APILOG('REQUEST', newObject.method, newUrl, newObject);

  const response = await fetch(newUrl, newObject);
  return handleResponse(response);
}
