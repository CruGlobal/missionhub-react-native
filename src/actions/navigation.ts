/* eslint-disable @typescript-eslint/no-explicit-any, import/named */

import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import {
  NavigationActions,
  StackActions,
  NavigationNavigateAction,
} from 'react-navigation';

import { FEED_ITEM_DETAIL_SCREEN } from '../containers/Communities/Community/CommunityFeedTab/FeedItemDetailScreen/FeedItemDetailScreen';
import { MAIN_TABS, PEOPLE_TAB, COMMUNITIES_TAB } from '../constants';
import { COMMUNITY_TABS } from '../containers/Communities/Community/constants';

import { loadHome } from './auth/userData';

export function navigatePush(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.push({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export function navigateBack(times?: number) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    if (times && times > 1) {
      dispatch(StackActions.pop({ n: times, immediate: true }));
    } else {
      dispatch(NavigationActions.back());
    }
  };
}

export const navigateReset = (screen: string, props = {}) =>
  resetStack([
    NavigationActions.navigate({ routeName: screen, params: props }),
  ]);

export const navigateNestedReset = (
  screens: {
    routeName: string;
    tabName?: string;
    params?: { [key: string]: any };
  }[],
) =>
  resetStack(
    screens.map(({ routeName, params, tabName }) =>
      NavigationActions.navigate({
        routeName,
        ...(params ? { params } : {}),
        ...(tabName
          ? {
              action: NavigationActions.navigate({
                routeName: tabName,
              }),
            }
          : {}),
      }),
    ),
    screens.length - 1,
  );

const resetStack = (actions: NavigationNavigateAction[], index = 0) => (
  dispatch: ThunkDispatch<{}, {}, AnyAction>,
) =>
  dispatch(
    StackActions.reset({
      index,
      key: null, // Reset root stack navigator
      actions,
    }),
  );

// The reset home and reset login are handled by the login/logout auth actions

export function navigateReplace(screen: string, props = {}) {
  return (dispatch: ThunkDispatch<{}, {}, AnyAction>) => {
    dispatch(
      StackActions.replace({
        routeName: screen,
        params: props,
      }),
    );
  };
}

export const navigateToMainTabs = (tabName = PEOPLE_TAB) => (
  dispatch: ThunkDispatch<never, {}, AnyAction>,
) => {
  dispatch(loadHome());
  dispatch(navigateNestedReset([{ routeName: MAIN_TABS, tabName }]));
};

export const navigateToFeedItemComments = (
  feedItemId: string,
  communityId: string,
) => (dispatch: ThunkDispatch<{}, null, AnyAction>) => {
  dispatch(
    navigateNestedReset([
      {
        routeName: MAIN_TABS,
        tabName: COMMUNITIES_TAB,
      },
      {
        routeName: COMMUNITY_TABS,
        params: { communityId },
      },
      {
        routeName: FEED_ITEM_DETAIL_SCREEN,
        params: { feedItemId },
      },
    ]),
  );
};
