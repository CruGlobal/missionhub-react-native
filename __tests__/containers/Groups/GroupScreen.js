import React from 'react';

import {
  GroupScreen,
  CRU_TABS,
  USER_CREATED_TABS,
} from '../../../src/containers/Groups/GroupScreen';
import {
  testSnapshotShallow,
  createMockNavState,
  createMockStore,
  renderShallow,
} from '../../../testUtils';
import { ADD_CONTACT_SCREEN } from '../../../src/containers/AddContactScreen';
import { navigatePush, navigateReset } from '../../../src/actions/navigation';
import { MAIN_TABS } from '../../../src/constants';
import * as common from '../../../src/utils/common';

jest.mock('../../../src/actions/navigation', () => ({
  navigateBack: jest.fn(() => ({ type: 'test' })),
  navigatePush: jest.fn(),
  navigateReset: jest.fn(),
}));

const organization = { id: '5', name: 'Test  Org', user_created: false };

describe('GroupScreen', () => {
  const createHeader = org => (
    <GroupScreen
      navigation={createMockNavState({
        organization: org,
      })}
    />
  );

  it('should render header correctly', () => {
    testSnapshotShallow(createHeader(organization));
  });

  it('should render header correctly for user_created org', () => {
    testSnapshotShallow(createHeader({ ...organization, user_created: true }));
  });

  it('should render Cru Community tabs correctly', () => {
    expect(CRU_TABS).toMatchSnapshot();
  });

  it('should render User Created Community tabs correctly', () => {
    expect(USER_CREATED_TABS).toMatchSnapshot();
  });

  it('should handle add contact button correctly', () => {
    const instance = renderShallow(
      <GroupScreen
        navigation={createMockNavState({
          organization,
        })}
        store={createMockStore()}
      />,
    ).instance();

    instance.handleAddContact();

    expect(navigatePush).toHaveBeenCalledWith(ADD_CONTACT_SCREEN, {
      onComplete: expect.anything(),
      organization,
    });
  });

  it('should handle go back correctly', () => {
    const component = renderShallow(
      <GroupScreen
        navigation={createMockNavState({
          organization,
        })}
        store={createMockStore()}
      />,
    );

    component.props().left.props.onPress();

    expect(navigateReset).toHaveBeenCalledWith(MAIN_TABS, {
      startTab: 'groups',
    });
  });

  it('calls disable back add', () => {
    const instance = renderShallow(
      <GroupScreen
        navigation={createMockNavState({
          organization,
        })}
        store={createMockStore()}
      />,
    ).instance();

    common.disableBack = { add: jest.fn() };
    instance.componentDidMount();
    expect(common.disableBack.add).toHaveBeenCalledTimes(1);
  });

  it('calls disable back remove', () => {
    const instance = renderShallow(
      <GroupScreen
        navigation={createMockNavState({
          organization,
        })}
        store={createMockStore()}
      />,
    ).instance();

    common.disableBack = { remove: jest.fn() };
    instance.componentWillUnmount();
    expect(common.disableBack.remove).toHaveBeenCalledTimes(1);
  });
});
