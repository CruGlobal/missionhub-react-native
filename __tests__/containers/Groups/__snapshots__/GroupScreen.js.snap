// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupScreen should render header correctly 1`] = `
<Connect(Header)
  left={<Connect(BackButton) />}
  right={
    <IconButton
      name="addContactIcon"
      onPress={[Function]}
      size={24}
      type="MissionHub"
    />
  }
  shadow={false}
  title="Test  Org"
/>
`;
